Docs BYOC content (#13976)
Lightning-AI/lightning: docs/source-app/workflows/byoc/index.rst

Next, as part of the environment creation process, you can configure networking,
security, and select among cluster configuration options based on their own use cases.

After submitting a cluster creation request, the Lightning Control Plane creates the required cloud infrastructure on the user account. This
sets up a new Lightning Cluster along with a Lightning Kubernetes Operator.

*******************************
Create a Lightning BYOC cluster
*******************************

You must have your cloud configured before you try and create a BYOC cluster.

Create a Lightning BYOC cluster using the following command:

.. code:: python

   lightning create cluster <cluster-name> <cloud-provider-parameters>

Here's an example:

.. code:: python

   lightning create cluster my-byoc-cluster --provider aws --role-arn arn:aws:iam::1234567890:role/lai-byoc --external-id dummy --region us-west-2 --instance-types t3.xlarge --cost-savings

..note:: Cluster creation is gonna take an hour or more after you run this command.

**Parameters:**

* cluster_name: The name of the cluster to be created

.. note:: Cluster names can only contain lowercase letters, numbers, and periodic hyphens ( - ).

> * provider: The cloud provider where your cluster is located.

Reviewer: @Felonious-Spellfire should we highlight flags?

`cluster_name` is an argument, but everything else needs a `--` prefix to work, ie `--provider`, `--external-id`, etc?

Author: I think because we provided the example people will understand what to do.

.. note:: AWS is supported today, but support for other cloud providers is coming soon.

> * role_arn: AWS IAM Role ARN used to provision resources

Reviewer: ```suggestion
* role-arn: AWS IAM Role ARN used to provision resources
```

> * external_id: AWS IAM Role external ID

Reviewer: ```suggestion
* external-id: AWS IAM Role external ID
```

.. note:: To read more on what the AWS external ID is and why it's useful go `here <https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html`_.

* region: AWS region containing compute resources

* instance_types: Instance types that you want to support, for computer jobs within the cluster. For now, this is the AWS instance types supported by the cluster.

> * cost_savings: Specifies if the cluster uses cost savings mode.

Reviewer: @Felonious-Spellfire this flag is being re-named to `enable-performance` in https://github.com/Lightning-AI/lightning/pull/14132

Author: You updated it. Thank you

> * cost_savings: Specifies if the cluster uses cost savings mode.

Reviewer: ```suggestion
* enable-performance: Specifies if the cluster uses cost savings mode.
```

.. note:: In cost saving mode the number of compute nodes is reduced to one, reducing the cost for clusters with low utilization.

> * edit_before_creation: Enables interactive editing of requests before submitting it to Lightning AI.

Reviewer: ```suggestion
* edit-before-creation: Enables interactive editing of requests before submitting it to Lightning AI.
```

see above

BYOC separates the control and data plane. The data plane, that includes
Lightning clusters, services and Lightning Apps, reside inside the user’s VPC.
The control plane resides on Lightning Cloud.

Setup begins with configuring a cloud provider (today AWS, but more are coming soon) with your personal credentials for
delegated access and an identity provider for secure access to the data plane.

Next, as part of the environment creation process, you can configure networking,
security, and select among cluster configuration options based on their own use cases.

After submitting a cluster creation request, the Lightning Control Plane creates the required cloud infrastructure on the user account. This
sets up a new Lightning Cluster along with a Lightning Kubernetes Operator.

*******************************
Create a Lightning BYOC cluster
*******************************

You must have your cloud configured before you try and create a BYOC cluster.

And to make your life a little easier, we've made a `Terraform module to help with that <https://github.com/Lightning-AI/terraform-aws-lightning-byoc>`_.

Create a Lightning BYOC cluster using the following command:

> .. code:: python

Reviewer 2: ```suggestion
.. code:: bash
```

   lightning create cluster <cluster-name> <cloud-provider-parameters>

Here's an example:

> .. code:: python

Reviewer 2: ```suggestion
.. code:: bash
```

   lightning create cluster my-byoc-cluster --provider aws --role-arn arn:aws:iam::1234567890:role/lai-byoc --external-id dummy --region us-west-2 --instance-types t3.xlarge --enable-performance

> ..note:: Cluster creation is gonna take an hour or more after you run this command.

Reviewer 2: I would expect that in the docs we always stay formal.

```suggestion
..note:: Cluster creation is going to take an hour or more after you run this command.
```

Author: @awaelchli The docs are not always formal. This was a choice made by William. Here's an example from this page: https://lightning.ai/lightning-docs/get_started/lightning_apps_intro.html

"
Run the app locally with the run command 🤯
"

If this is allowed, then informal conversational writing should also be acceptable.

**Arguments**

* cluster_name: The name of the cluster to be created

.. note:: Cluster names can only contain lowercase letters, numbers, and periodic hyphens ( - ).

**Parameters:**

* provider: The cloud provider where your cluster is located.

.. note:: AWS is supported today, but support for other cloud providers is coming soon.

* role-arn: AWS IAM Role ARN used to provision resources

* external-id: AWS IAM Role external ID

.. note:: To read more on what the AWS external ID is and why it's useful go `here <https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html`_.

* region: AWS region containing compute resources

* instance-types: Instance types that you want to support, for computer jobs within the cluster. For now, this is the AWS instance types supported by the cluster.

* enable-performance: Specifies if the cluster uses cost savings mode.

> .. note:: In cost saving mode the number of compute nodes is reduced to one, reducing the cost for clusters with low utilization.

Reviewer 2: Note: too many of these notes will disturb the reader visually, because it will render a big blue box around it. I suggest to minimize this usage. In a bullet point list, you may also need to add indentation to make it render properly.

Author: So the trouble is, we want to add help at the point where it is needed. I could make this a table, but the notes will still be included, so I'm not sure how much ahead we will be.

Author: I've put the content into a table. It's a bit better.

* edit-before-creation: Enables interactive editing of requests before submitting it to Lightning AI.

* wait: Waits for the cluster to be in a RUNNING state. Only use this for debugging.

----

*******************************************
View a list of your Lightning BYOC clusters
*******************************************

> .. code:: python

Reviewer 2: ```suggestion
.. code:: bash
```

   lightning list clusters

> *******************************

Reviewer 2: why is this section grouped without the top one? Shouldn't there be one of these bug spaces? (----)

Author: Updated

Delete a Lightning BYOC cluster
*******************************

Deletes a Lightning BYOC cluster. Lightning AI removes cluster artifacts and any resources running on the cluster.

.. warning:: Using the --force parameter when deleting a cluster does not clean up any resources managed by Lightning AI. Check your cloud provider to verify that existing cloud resources are deleted.

Deletion permanently removes not only the record of all runs on a cluster, but all associated artifacts, metrics, logs, etc.

.. warning:: This process may take a few minutes to complete, but once started it CANNOT be rolled back. Deletion permanently removes not only the BYOC cluster from being managed by Lightning AI, but tears down every BYOC resource Lightning AI managed (for that cluster id) in the host cloud. All object stores, container registries, logs, compute nodes, volumes, etc. are deleted and cannot be recovered.

> .. code:: python

Reviewer 2: ```suggestion
.. code:: bash
```

#################################
Run Apps on your own cloud (BYOC)
#################################

**Audience:** Users looking to run Lightning Apps on their own private cloud.

----

*******************
A bit of background
*******************

> BYOC - Bring Your Own Cloud, is an alternate deployment model to Lightning Cloud (fully managed SaaS).

Reviewer: nit: I think we use `BYOC` as "Bring your own credentials"

Author: Will verify terms

#################################
Run Apps on your own cloud (BYOC)
#################################

**Audience:** Users looking to run Lightning Apps on their own private cloud.

----

*******************
A bit of background
*******************

BYOC - Bring Your Own Cloud, is an alternate deployment model to Lightning Cloud (fully managed SaaS).
BYOC separates the control and data plane. The data plane, that includes
Lightning clusters, services and Lightning Apps, reside inside the user’s VPC.

> The control plane resides on Lightning Cloud.

Reviewer 3: ```suggestion
The control plane resides on Lightning Cloud.
```

> Setup begins with configuring a cloud provider (today AWS, but more are coming soon) with credentials for

Reviewer 3: ```suggestion
Setup begins with configuring a cloud provider (today AWS, but more are coming soon) with your personal credentials for
```

delegated access and an identity provider for secure access to the data plane.

> Next, as part of the environment creation process, users can configure networking,

Reviewer 4: Are we using firt person?

Author: Good catch. Updated

> Next, as part of the environment creation process, users configure networking,
> security, and select Lightning cluster configuration options based on their use cases.
> 
> After submitting the request,the Lightning control plane creates the required cloud infrastructure
> and sets up the environment with the Lightning cluster along with a Lightning Kubernetes operator.

Reviewer 3: ```suggestion
Next, as part of the environment creation process, users can configure networking,
security, and select among cluster configuration options based on their own use cases.

After submitting a cluster creation request, the Lightning Control Plane creates the required cloud infrastructure on the user account. This
sets up a new Lightning Cluster along with a Lightning Kubernetes Operator.
```

*******************************
Create a Lightning BYOC cluster

> *******************************

Reviewer: I think we need a section here noting that cluster creation can take a long time (1+hrs)

Author: Updated.

You must have your cloud configured before you try and create a BYOC cluster.

Create a Lightning BYOC cluster using the following command:

> PLEASE PROVIDE AN EXAMPLE WITH CLOUD PROVIDER PARAMETERS

Reviewer 3: @edenlightning

Reviewer: here's an example request to create a cluster:

```
lightning create cluster test-raphael9 --provider aws --role-arn arn:aws:iam::1234567890:role/lai-byoc --external-id dummy --region us-west-2 --instance-types t3.xlarge --cost-savings
```

a full list of parameters is available [here](https://github.com/Lightning-AI/lightning/blob/master/src/lightning_app/cli/lightning_cli_create.py#L13)

I think we should link to our [terraform bootstrap](https://github.com/Lightning-AI/terraform-aws-lightning-byoc) repository as well - just need to make it public first.

Author: Added something for the Terraform module. Please take a peek

.. note:: AWS is supported today, but support for other cloud providers is coming soon.

Args:

* cost_savings: Specifies if the cluster uses cost savings mode.

.. note:: In cost saving mode the number of compute nodes is reduced to one, reducing the cost for clusters with low utilization.

* cluster_name: The name of the cluster to be created

* role_arn: AWS IAM Role ARN used to provision resources

* region: AWS region containing compute resources

* external_id: AWS IAM Role external ID

* instance_types: AWS instance types supported by the cluster

* edit_before_creation: Enables interactive editing of requests before submitting it to Lightning AI.

* wait: Waits for the cluster to be in a RUNNING state. Only use this for debugging.

----

*******************************************
View a list of your Lightning BYOC clusters
*******************************************

.. code:: python

   lightning list clusters

*******************************
Delete a Lightning BYOC cluster
*******************************

Deletes a Lightning BYOC cluster. Lightning AI removes cluster artifacts and any resources running on the cluster.

> .. warning:: Deleting a cluster does not clean up any resources managed by Lightning AI. Check your cloud provider to verify that existing cloud resources are deleted.

Reviewer: it actually does, unless you pass the `--force` flag.

If you do not pass `--force` we'll delete everything but storage artifacts in S3 if I recall correctly.

Author: Updated. Please take a peek.

> Deleting a run also deletes all Runs and Experiments that were started on the cluster.

Reviewer: I think this needs re-wording as we don't support Runs or Experiments on Lightning?

Author: Removed this and other content that referred to experiments.

Deletion permanently removes not only the record of all runs on a cluster, but all associated experiments, artifacts, metrics, logs, etc.

.. warning:: This process may take a few minutes to complete, but once started it CANNOT be rolled back. Deletion permanently removes not only the BYOC cluster from being managed by Lightning AI, but tears down every BYOC resource Lightning AI managed (for that cluster id) in the host cloud. All object stores, container registries, logs, compute nodes, volumes, etc. are deleted and cannot be recovered.

.. code:: python

> 

Reviewer: can we include the `--force` flag here?

Author: Included it in the previous warning. Please take a peek.

sets up a new Lightning Cluster along with a Lightning Kubernetes Operator.

*******************************
Create a Lightning BYOC cluster
*******************************

You must have your cloud configured before you try and create a BYOC cluster.

Create a Lightning BYOC cluster using the following command:

.. code:: python

   lightning create cluster <cluster-name> <cloud-provider-parameters>

Here's an example:

.. code:: python

   lightning create cluster my-byoc-cluster --provider aws --role-arn arn:aws:iam::1234567890:role/lai-byoc --external-id dummy --region us-west-2 --instance-types t3.xlarge --cost-savings

..note:: Cluster creation is gonna take an hour or more after you run this command. 

**Parameters:**

* cluster_name: The name of the cluster to be created

.. note:: Cluster names can only contain lowercase letters, numbers, and periodic hyphens ( - ).

* provider: The cloud provider where your cluster is located.

.. note:: AWS is supported today, but support for other cloud providers is coming soon.

* role_arn: AWS IAM Role ARN used to provision resources

> * external_id: AWS IAM Role external ID

Reviewer: based on wills feedback, consider linking to https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html to explain what an external ID is and why it's useful.

Author: Updated

Author: Updated
